rust/bridge/node/src/lib.rs

```diff
@@ -3,13 +3,287 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "callback error {}", self.message)
+    }
+}
+
+impl std::error::Error for CallbackError {}
+
+fn js_error_to_rust(func: &'static str, err: String) -> SignalProtocolError {
+    SignalProtocolError::ApplicationCallbackError(func, Box::new(CallbackError::new(err)))
+}
+
+pub fn boxed_object<'a, T: 'static + Send>(
+    cx: &mut TaskContext<'a>,
+    value: T,
+) -> JsResult<'a, JsValue> {
+    Ok(cx.boxed(node::DefaultFinalize(value)).upcast())
+}
+
+struct NodeSenderKeyStore<'a> {
+    js_queue: EventQueue,
+    store_object: Arc<Root<JsObject>>,
+    phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> NodeSenderKeyStore<'a> {
+    fn new(cx: &mut FunctionContext, store: Handle<JsObject>) -> Self {
+        Self {
+            js_queue: cx.queue(),
+            store_object: Arc::new(store.root(cx)),
+            phantom: PhantomData,
+        }
+    }
+
+    async fn do_get_sender_key(
+        &self,
+        name: SenderKeyName,
+    ) -> Result<Option<SenderKeyRecord>, String> {
+        let store_object_shared = self.store_object.clone();
+        JsFuture::get_promise(&self.js_queue, move |cx| {
+            let store_object = store_object_shared.to_inner(cx);
+            let name: Handle<JsValue> = boxed_object(cx, name)?;
+            let result = call_method(cx, store_object, "_getSenderKey", std::iter::once(name))?;
+            let result = result.downcast_or_throw(cx)?;
+            store_object_shared.finalize(cx);
+            Ok(result)
+        })
+        .then(|cx, result| match result {
+            Ok(value) => {
+                match value.downcast::<JsBox<crate::node::DefaultFinalize<SenderKeyRecord>>, _>(cx)
+                {
+                    Ok(obj) => Ok(Some((***obj).clone())),
+                    Err(_) => {
+                        if value.is_a::<JsNull, _>(cx) {
+                            Ok(None)
+                        } else {
+                            Err("result must be an object".into())
+                        }
+                    }
+                }
+            }
+            Err(error) => Err(error
+                .to_string(cx)
+                .expect("can convert to string")
+                .value(cx)),
+        })
+        .await
+    }
+
+    async fn do_save_sender_key(
+        &self,
+        name: SenderKeyName,
+        record: SenderKeyRecord,
+    ) -> Result<(), String> {
+        let store_object_shared = self.store_object.clone();
+        JsFuture::get_promise(&self.js_queue, move |cx| {
+            let store_object = store_object_shared.to_inner(cx);
+            let name: Handle<JsValue> = boxed_object(cx, name)?;
+            let record: Handle<JsValue> = boxed_object(cx, record)?;
+            let result = call_method(
+                cx,
+                store_object,
+                "_saveSenderKey",
+                std::iter::once(name).chain(std::iter::once(record)),
```

Comment:
I think `vec!` is fine here; worrying about a small heap allocation when we're going to call into JavaScript is probably overkill.